Libraries/LibWeb/XPath/XPath.cpp

```diff
@@ -0,0 +1,223 @@
+    }
+    case DOM::NodeType::CDATA_SECTION_NODE: {
+        auto const& cdata = static_cast<DOM::CDATASection const&>(node);
+        ByteString data = cdata.data().to_byte_string();
+        auto* xml_cdata = xmlNewCDataBlock(doc, reinterpret_cast<xmlChar const*>(data.characters()), data.length());
+        xml_cdata->_private = bit_cast<void*>(&node);
+        return xml_cdata;
+    }
+    case DOM::NodeType::ENTITY_REFERENCE_NODE: {
+        TODO(); // Does not seem to be used at all in ladybird
+    }
+    case DOM::NodeType::ENTITY_NODE: {
+        TODO(); // Entity nodes are unused in libxml2
+    }
+    case DOM::NodeType::PROCESSING_INSTRUCTION_NODE: {
+        auto const& processing_instruction = static_cast<DOM::ProcessingInstruction const&>(node);
+        auto* xml_pi = xmlNewDocPI(doc, reinterpret_cast<xmlChar const*>(processing_instruction.target().to_byte_string().characters()), reinterpret_cast<xmlChar const*>(processing_instruction.data().to_byte_string().characters()));
+        xml_pi->_private = bit_cast<void*>(&node);
+        return xml_pi;
+    }
+    case DOM::NodeType::COMMENT_NODE: {
+        auto const& comment = static_cast<DOM::Comment const&>(node);
+        auto* xml_comment = xmlNewDocComment(doc, reinterpret_cast<xmlChar const*>(comment.data().to_byte_string().characters()));
+        xml_comment->_private = bit_cast<void*>(&node);
+        return xml_comment;
+    }
+    case DOM::NodeType::DOCUMENT_NODE: {
+        auto const& document = static_cast<DOM::Document const&>(node);
+        return mirror_node(doc, *document.document_element());
+    }
+    case DOM::NodeType::DOCUMENT_TYPE_NODE: {
+        TODO(); // Unused in libxml2
+    }
+    case DOM::NodeType::DOCUMENT_FRAGMENT_NODE: {
+        auto const& fragment = static_cast<DOM::DocumentFragment const&>(node);
+        auto* xml_fragment = xmlNewDocFragment(doc);
+        xml_fragment->_private = bit_cast<void*>(&node);
+        auto children = fragment.children_as_vector();
+        for (auto& child : children) {
+            xmlAddChild(xml_fragment, mirror_node(doc, *child));
+        }
+        return xml_fragment;
+    }
+    case DOM::NodeType::NOTATION_NODE: {
+        TODO(); // Unused in libxml2
+    }
+    }
+    return nullptr;
+}
+
+void convert_xpath_result(xmlXPathObjectPtr xpath_result, XPath::XPathResult* result, unsigned short type)
+{
+    if (!xpath_result) {
+        return;
+    }
+
+    switch (xpath_result->type) {
+    case XPATH_UNDEFINED:
+        break;
+    case XPATH_NODESET: {
+        Vector<GC::Ptr<DOM::Node>> node_list;
+
+        if (xpath_result->nodesetval && xpath_result->nodesetval->nodeNr > 0) {
+            for (int i = 0; i < xpath_result->nodesetval->nodeNr; i++) {
+                auto* node = xpath_result->nodesetval->nodeTab[i];
+                auto* dom_node = static_cast<DOM::Node*>(node->_private);
+
+                GC::Ptr<DOM::Node> ref(*dom_node);
+
+                node_list.append(ref);
```

Comment:
You should be able to simply append `dom_node` here as a pointer. Make it `unchecked_append` if you added `ensure_capacity` above.